File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/ThrowKeywordRecommender.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Threading;
using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;

namespace Microsoft.CodeAnalysis.CSharp.Completion.KeywordRecommenders
{
    internal class ThrowKeywordRecommender : AbstractSyntacticSingleKeywordRecommender
    {
        public ThrowKeywordRecommender()
            : base(SyntaxKind.ThrowKeyword)
        {
        }

        protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
        {
            if (context.IsStatementContext || context.IsGlobalStatementContext)
            {
                return true;
            }

            // void M() => throw
            if (context.TargetToken.IsKind(SyntaxKind.EqualsGreaterThanToken))
            {
                return true;
            }

            // val ?? throw
            if (context.TargetToken.IsKind(SyntaxKind.QuestionQuestionToken))
            {
                return true;
            }

            //  expr ? throw : ...
            //  expr ? ... : throw
            if (context.TargetToken.Kind() is SyntaxKind.QuestionToken or
                SyntaxKind.ColonToken)
            {
                return context.TargetToken.Parent.IsKind(SyntaxKind.ConditionalExpression);
            }

            return false;
        }
    }
}

File: src/Features/VisualBasic/Portable/Completion/KeywordRecommenders/Statements/ThrowKeywordRecommender.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.Threading
Imports Microsoft.CodeAnalysis.Completion.Providers
Imports Microsoft.CodeAnalysis.VisualBasic.Extensions.ContextQuery

Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.KeywordRecommenders.Statements
    ''' <summary>
    ''' Recommends the "Throw" keyword for the statement context
    ''' </summary>
    Friend Class ThrowKeywordRecommender
        Inherits AbstractKeywordRecommender

        Protected Overrides Function RecommendKeywords(context As VisualBasicSyntaxContext, cancellationToken As CancellationToken) As IEnumerable(Of RecommendedKeyword)
            If context.IsSingleLineStatementContext Then
                Return SpecializedCollections.SingletonEnumerable(New RecommendedKeyword("Throw", VBFeaturesResources.Throws_an_exception_within_a_procedure_so_that_you_can_handle_it_with_structured_or_unstructured_exception_handling_code))
            End If

            Return SpecializedCollections.EmptyEnumerable(Of RecommendedKeyword)()
        End Function
    End Class
End Namespace
